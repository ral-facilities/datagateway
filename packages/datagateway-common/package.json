{
  "name": "datagateway-common",
  "version": "0.1.0",
  "private": true,
  "files": [
    "lib"
  ],
  "main": "./lib/index.js",
  "dependencies": {
    "@date-io/date-fns": "^1.3.13",
    "@material-ui/pickers": "^3.2.10",
    "@material-ui/lab": "^4.0.0-alpha.57",
    "@types/lodash.debounce": "^4.0.6",
    "axios": "^0.21.1",
    "clsx": "^1.1.1",
    "connected-react-router": "^6.9.1",
    "date-fns": "^2.17.0",
    "i18next": "^19.9.0",
    "lodash.debounce": "^4.0.8",
    "loglevel": "^1.7.1",
    "react-draggable": "^4.4.3",
    "react-i18next": "^11.8.8",
    "react-redux": "^7.1.0",
    "react-router": "^5.0.1",
    "react-scripts": "4.0.3",
    "react-virtualized": "^9.22.3",
    "redux": "^4.0.4",
    "redux-mock-store": "^1.5.4",
    "redux-thunk": "^2.3.0"
  },
  "peerDependencies": {
    "@material-ui/core": ">=  4.0.0 < 5",
    "@material-ui/icons": ">=  4.0.0 < 5",
    "react": ">= 16.8.0 < 17",
    "react-dom": ">= 16.8.0 < 17",
    "react-router-dom": ">= 5.0.0 < 6"
  },
  "devDependencies": {
    "@material-ui/core": "^4.11.3",
    "@material-ui/icons": "^4.11.2",
    "@types/enzyme": "^3.10.8",
    "@types/jest": "^26.0.4",
    "@types/node": "^14.14.31",
    "@types/react": "^17.0.2",
    "@types/react-router-dom": "^5.1.7",
    "@types/react-virtualized": "^9.21.10",
    "@typescript-eslint/eslint-plugin": "^4.16.1",
    "@typescript-eslint/parser": "^4.16.1",
    "babel-eslint": "10.1.0",
    "enzyme": "^3.10.0",
    "enzyme-adapter-react-16": "^1.15.6",
    "enzyme-to-json": "^3.6.1",
    "eslint": "^7.5.0",
    "eslint-config-prettier": "^8.1.0",
    "eslint-config-react-app": "^6.0.0",
    "eslint-plugin-cypress": "^2.11.1",
    "eslint-plugin-flowtype": "^5.8.0",
    "eslint-plugin-import": "^2.22.1",
    "eslint-plugin-jsx-a11y": "^6.3.1",
    "eslint-plugin-prettier": "^3.3.1",
    "eslint-plugin-react": "^7.20.3",
    "eslint-plugin-react-hooks": "^4.1.0",
    "lint-staged": "^10.5.4",
    "prettier": "^2.2.1",
    "react": "^16.13.1",
    "react-dom": "^16.11.0",
    "react-router-dom": "^5.0.1",
    "typescript": "4.2.2",
    "tslib": "^2.1.0"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test --env=jsdom --coverage --watchAll=false",
    "prepare": "yarn tsc",
    "test:watch": "react-scripts test --env=jsdom --watch",
    "lint:js": "eslint --ext=tsx --ext=ts --ext=js --ext=jsx --fix ./src",
    "tsc": "tsc -p tsconfig.lib.json",
    "eject": "react-scripts eject",
    "pre-commit": "lint-staged"
  },
  "lint-staged": {
    "src/**/*.{tsx,ts,js,jsx,json}": [
      "eslint --max-warnings=0 --ext=tsx --ext=ts --ext=js --ext=jsx  --fix",
      "prettier --write"
    ],
    "cypress/**/*.{tsx,ts,js,jsx,json}": [
      "eslint --max-warnings=0 --ext=tsx --ext=ts --ext=js --ext=jsx --fix",
      "prettier --write"
    ]
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ]
  },
  "jest": {
    "snapshotSerializers": [
      "enzyme-to-json/serializer"
    ],
    "collectCoverageFrom": [
      "src/**/*.{tsx,ts,js,jsx}",
      "!src/index.tsx",
      "!src/serviceWorker.ts",
      "!src/setupTests.js"
    ]
  }
}
