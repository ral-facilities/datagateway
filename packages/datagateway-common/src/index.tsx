// Uncomment to use App to test components
// import React from 'react';
// import ReactDOM from 'react-dom';
// import App from './App';
// import './index.css';

import { StateType } from './state/app.types';

export * from './app.types';
export { default as Preloader } from './preloader/preloader.component';
export * from './state/actions/actions.types';

export { default as Table } from './table/table.component';
export * from './table/table.component';
export { default as DetailsPanelRow } from './table/rowRenderers/detailsPanelRow.component';
export { default as DataHeader } from './table/headerRenderers/dataHeader.component';
export {
  default as TextColumnFilter,
  useTextFilter,
  usePrincipalExperimenterFilter,
} from './table/columnFilters/textColumnFilter.component';
export {
  default as DateColumnFilter,
  useDateFilter,
} from './table/columnFilters/dateColumnFilter.component';
export { default as ActionCellComponent } from './table/cellRenderers/actionCell.component';
export { default as DataCellComponent } from './table/cellRenderers/dataCell.component';
export { default as ExpandCellComponent } from './table/cellRenderers/expandCell.component';
export * from './table/cellRenderers/cellContentRenderers';

export { default as CardView } from './card/cardView.component';
export type { CardViewDetails } from './card/cardView.component';
export { default as AdvancedFilter } from './card/advancedFilter.component';

export * from './state/actions/index';
export * from './state/middleware/dgcommon.middleware';
export * from './state/app.types';
export * from './api/index';

export { default as DGCommonMiddleware } from './state/middleware/dgcommon.middleware';
export { default as createReducer } from './state/reducers/createReducer';
export {
  default as dGCommonReducer,
  initialState as dGCommonInitialState,
} from './state/reducers/dgcommon.reducer';

export type DGCommonState = StateType;

export * from './parseTokens';
export { default as handleICATError } from './handleICATError';
export { default as retryICATErrors } from './api/retryICATErrors';

export {
  default as ArrowTooltip,
  getTooltipText,
} from './arrowtooltip.component';
export { default as Sticky } from './sticky.component';
export { default as DGThemeProvider } from './dgThemeProvider.component';
export { default as Mark } from './mark.component';

export { default as HomePage } from './homePage/homePage.component';

export { default as AddToCartButton } from './views/addToCartButton.component';
export { default as ViewCartButton } from './views/viewCartButton.component';
export type { CartProps } from './views/viewCartButton.component';
export { default as ViewButton } from './views/viewButton.component';
export { default as ClearFiltersButton } from './views/clearFiltersButton.component';
export { default as DownloadButton } from './views/downloadButton.component';
export { default as SelectionAlert } from './views/selectionAlert.component';

export { default as ISISDatafileDetailsPanel } from './detailsPanels/isis/datafileDetailsPanel.component';
export { default as ISISDatasetDetailsPanel } from './detailsPanels/isis/datasetDetailsPanel.component';
export { default as ISISInstrumentDetailsPanel } from './detailsPanels/isis/instrumentDetailsPanel.component';
export { default as ISISInvestigationDetailsPanel } from './detailsPanels/isis/investigationDetailsPanel.component';
export { default as DLSDatafileDetailsPanel } from './detailsPanels/dls/datafileDetailsPanel.component';
export { default as DLSDatasetDetailsPanel } from './detailsPanels/dls/datasetDetailsPanel.component';
export { default as DLSVisitDetailsPanel } from './detailsPanels/dls/visitDetailsPanel.component';
export { default as InvestigationDetailsPanel } from './detailsPanels/investigationDetailsPanel.component';
export { default as DatasetDetailsPanel } from './detailsPanels/datasetDetailsPanel.component';
export { default as DatafileDetailsPanel } from './detailsPanels/datafileDetailsPanel.component';

export * from './urlBuilders';

// ReactDOM.render(<App />, document.getElementById('root'));
