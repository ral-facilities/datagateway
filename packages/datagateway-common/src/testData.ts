import type {
  Datafile,
  Dataset,
  FacilityCycle,
  Investigation,
} from './app.types';

export const mockApiUrl = 'https://example.com/api';
export const mockFacilityName = 'LILS';

export const mockFacilityCycles: FacilityCycle[] = [
  {
    id: 12938,
    name: 'studio toughness',
    description: 'He decided water-skiing on a frozen lake wasnâ€™t a good idea.',
    startDate: '2006-01-20T16:30:17Z',
    endDate: '2007-01-20T16:30:17Z',
  },
  {
    id: 402,
    name: 'within cell interlinked',
    description: 'He waited for the stop sign to turn to a go sign.',
    startDate: '2017-03-17T14:03:11Z',
    endDate: '2020-11-29T05:41:54Z',
  },
];

export const mockInvestigations: Investigation[] = [
  {
    id: 58,
    title: 'Happiness can be found in the depths of chocolate pudding.',
    name: 'investigation news',
    visitId: 'CqJN',
    startDate: '2018-03-09T08:19:55Z',
    endDate: '2018-03-29T08:19:55Z',
    investigationInstruments: [
      {
        id: 446,
        instrument: {
          id: 937,
          name: 'instrument fame',
        },
      },
    ],
    investigationFacilityCycles: [
      {
        id: 23,
        facilityCycle: {
          id: 402,
          name: 'within cell interlinked',
          description: 'He waited for the stop sign to turn to a go sign.',
          startDate: '2017-03-17T14:03:11Z',
          endDate: '2020-11-29T05:41:54Z',
        },
      },
    ],
  },
  {
    id: 993,
    title: 'I ate a sock because people on the Internet told me to.',
    name: 'investigation inn',
    visitId: 'z0bLi1f3',
    startDate: '2019-08-08T22:27:07Z',
    endDate: '2019-09-08T22:27:07Z',
    investigationInstruments: [
      {
        id: 262,
        instrument: {
          id: 927,
          name: 'instrument case',
        },
      },
    ],
    investigationFacilityCycles: [],
  },
];

export const mockDatasets: Dataset[] = [
  {
    id: 856,
    name: 'dataset modern',
    investigation: mockInvestigations[0],
    createTime: '2018-03-10T08:19:55Z',
    modTime: '2018-03-10T09:19:55Z',
  },
  {
    id: 535,
    name: 'dataset lazy',
    investigation: mockInvestigations[1],
    createTime: '2019-09-01T22:27:07Z',
    modTime: '2019-09-02T22:27:07Z',
  },
];

export const mockDatafiles: Datafile[] = [
  {
    id: 70,
    name: 'datafile weekend',
    modTime: '2018-03-10T08:19:55Z',
    createTime: '2018-03-10T08:19:55Z',
    dataset: mockDatasets[0],
  },
];
