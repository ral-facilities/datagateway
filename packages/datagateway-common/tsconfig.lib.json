{
  "compilerOptions": {
    "baseUrl": ".",
    "outDir": "./lib",
    "target": "es5",
    "sourceMap": true,
    "rootDir": "src",
    "declaration": true,
    "declarationMap": true,
    "noImplicitReturns": true,
    "noImplicitThis": true,
    "noImplicitAny": true,
    "importHelpers": true,
    "strictNullChecks": true,
    "suppressImplicitAnyIndexErrors": true,
    "noUnusedLocals": false,
    "lib": ["dom", "dom.iterable", "esnext"],
    "skipLibCheck": true,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "jsx": "react"
  },
  "exclude": [
    "lib",
    "node_modules",
    "build",
    "scripts",
    "acceptance-tests",
    "webpack",
    "jest",
    "src/setupTests.tsx",
    "**/*.test.*"
  ],
  "include": ["src"]
}
