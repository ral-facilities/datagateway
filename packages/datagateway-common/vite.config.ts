import react from '@vitejs/plugin-react';
// import browserslistToEsbuild from 'browserslist-to-esbuild';
import fs from 'node:fs';
import path from 'path';
import url from 'url';
import { PluginOption, UserConfig, defineConfig, loadEnv } from 'vite';
import browserslistToEsbuild from 'browserslist-to-esbuild';

// see https://github.com/bvaughn/react-virtualized/issues/1722#issuecomment-1893242667
function reactVirtualized(): PluginOption {
  const WRONG_CODE = `import { bpfrpt_proptype_WindowScroller } from "../WindowScroller.js";`;

  return {
    name: 'my:react-virtualized',
    async configResolved() {
      const reactVirtualizedPath = path.dirname(
        url.fileURLToPath(import.meta.resolve('react-virtualized'))
      );

      const brokenFilePath = path.join(
        reactVirtualizedPath,
        '..', // back to dist
        'es',
        'WindowScroller',
        'utils',
        'onScroll.js'
      );
      const brokenCode = await fs.promises.readFile(brokenFilePath, 'utf-8');

      const fixedCode = brokenCode.replace(WRONG_CODE, '');
      await fs.promises.writeFile(brokenFilePath, fixedCode);
    },
  };
}

// Obtain default coverage config from vitest when not building for production
// (to avoid importing vitest during build as its a dev dependency)
let vitestDefaultExclude: string[] = [];
let vitestCoverageConfigDefaultsExclude: string[] = [];
if (process.env.NODE_ENV !== 'production') {
  await import('vitest/config').then((vitestConfig) => {
    vitestDefaultExclude = vitestConfig.defaultExclude;
    vitestCoverageConfigDefaultsExclude =
      vitestConfig.coverageConfigDefaults.exclude;
  });
}

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => {
  const env = loadEnv(mode, process.cwd(), '');

  const plugins: PluginOption[] = [react(), reactVirtualized()];

  const config: UserConfig = {
    plugins: plugins,
    server: {
      port: 3000,
    },
    preview: {
      port: 5001,
    },
    define: {
      // See https://vitejs.dev/guide/build.html#library-mode
      // we need to replace here as the build in library mode won't
      'process.env.NODE_ENV': JSON.stringify(env.NODE_ENV),
    },
  };

  config.build = {
    lib: {
      entry: 'src/main.tsx',
      name: 'datagateway-common',
    },
  };

  config.build.target = browserslistToEsbuild();

  return {
    ...config,
    test: {
      globals: true,
      environment: 'jsdom',
      setupFiles: ['src/setupTests.tsx'],
      exclude: [...vitestDefaultExclude, 'e2e/**'],
      coverage: {
        reporter: [
          // Default
          'text',
          'html',
          'clover',
          'json',
          // Extra for VSCode extension
          ['lcov', { outputFile: 'lcov.info', silent: true }],
        ],
        exclude: [
          ...vitestCoverageConfigDefaultsExclude,
          'public/*',
          'src/vite-env.d.ts',
          'src/main.tsx',
        ],
      },
    },
  };
});
