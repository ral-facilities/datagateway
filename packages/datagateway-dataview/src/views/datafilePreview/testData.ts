// This file contains mock data for testing datafile preview functionality.

import type { Datafile } from 'datagateway-common';

const mockDatafile: Datafile = {
  createTime: '2020-01-01',
  dataset: undefined,
  description: 'test description',
  fileSize: 100,
  id: 123,
  location: 'test location',
  modTime: '2020-01-02',
  name: 'Datafile.txt',
  parameters: [],
};

const mockTxtFileContent = `First line
Second line
Third line
`;

export { mockDatafile, mockTxtFileContent };
