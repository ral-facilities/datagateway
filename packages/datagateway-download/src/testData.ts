import {
  Download,
  DownloadCartItem,
  FormattedDownload,
} from 'datagateway-common';
import type { DownloadSettings } from './ConfigProvider';

export const mockDownloadItems: Download[] = [
  {
    createdAt: '2020-02-25T15:05:29Z',
    downloadItems: [{ entityId: 1, entityType: 'investigation', id: 1 }],
    email: 'test1@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-1',
    fullName: 'Person 1',
    id: 1,
    isDeleted: false,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 1000,
    status: 'COMPLETE',
    transport: 'https',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-26T15:05:35Z',
    downloadItems: [{ entityId: 2, entityType: 'investigation', id: 2 }],
    email: 'test2@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-2',
    fullName: 'Person 2',
    id: 2,
    isDeleted: false,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 2000,
    status: 'PREPARING',
    transport: 'globus',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-27T15:57:20Z',
    downloadItems: [{ entityId: 3, entityType: 'investigation', id: 3 }],
    email: 'test3@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-3',
    fullName: 'Person 3',
    id: 3,
    isDeleted: false,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 3000,
    status: 'RESTORING',
    transport: 'https',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-28T15:57:28Z',
    downloadItems: [{ entityId: 4, entityType: 'investigation', id: 4 }],
    email: 'test4@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-4',
    fullName: 'Person 4',
    id: 4,
    isDeleted: true,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 4000,
    status: 'EXPIRED',
    transport: 'globus',
    userName: 'test user',
  },
  {
    createdAt: '2020-03-01T15:57:28Z[UTC]',
    downloadItems: [{ entityId: 5, entityType: 'investigation', id: 5 }],
    email: 'test5@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-5',
    fullName: 'Person 5',
    id: 5,
    isDeleted: false,
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 5000,
    status: 'PAUSED',
    transport: 'globus',
    userName: 'test user',
  },
];

export const mockFormattedDownloadItems: FormattedDownload[] = [
  {
    createdAt: '2020-02-25T15:05:29Z',
    downloadItems: [{ entityId: 1, entityType: 'investigation', id: 1 }],
    email: 'test1@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-1',
    fullName: 'Person 1',
    id: 1,
    isDeleted: false,
    formattedIsDeleted: 'No',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 1000,
    status: 'COMPLETE',
    formattedStatus: 'downloadStatus.complete',
    transport: 'https',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-26T15:05:35Z',
    downloadItems: [{ entityId: 2, entityType: 'investigation', id: 2 }],
    email: 'test2@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-2',
    fullName: 'Person 2',
    id: 2,
    isDeleted: false,
    formattedIsDeleted: 'No',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 2000,
    status: 'PREPARING',
    formattedStatus: 'downloadStatus.preparing',
    transport: 'globus',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-27T15:57:20Z',
    downloadItems: [{ entityId: 3, entityType: 'investigation', id: 3 }],
    email: 'test3@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-3',
    fullName: 'Person 3',
    id: 3,
    isDeleted: false,
    formattedIsDeleted: 'No',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 3000,
    status: 'RESTORING',
    formattedStatus: 'downloadStatus.restoring',
    transport: 'https',
    userName: 'test user',
  },
  {
    createdAt: '2020-02-28T15:57:28Z',
    downloadItems: [{ entityId: 4, entityType: 'investigation', id: 4 }],
    email: 'test4@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-4',
    fullName: 'Person 4',
    id: 4,
    isDeleted: true,
    formattedIsDeleted: 'Yes',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 4000,
    status: 'EXPIRED',
    formattedStatus: 'downloadStatus.expired',
    transport: 'globus',
    userName: 'test user',
  },
  {
    createdAt: '2020-03-01T15:57:28Z[UTC]',
    downloadItems: [{ entityId: 5, entityType: 'investigation', id: 5 }],
    email: 'test5@email.com',
    facilityName: 'LILS',
    fileName: 'test-file-5',
    fullName: 'Person 5',
    id: 5,
    isDeleted: false,
    formattedIsDeleted: 'No',
    isEmailSent: true,
    isTwoLevel: false,
    preparedId: 'test-prepared-id',
    sessionId: 'test-session-id',
    size: 5000,
    status: 'PAUSED',
    formattedStatus: 'downloadStatus.paused',
    transport: 'globus',
    userName: 'test user',
  },
];

export const mockCartItems: DownloadCartItem[] = [
  {
    entityId: 1,
    entityType: 'investigation',
    id: 1,
    name: 'INVESTIGATION 1',
    parentEntities: [],
  },
  {
    entityId: 2,
    entityType: 'investigation',
    id: 2,
    name: 'INVESTIGATION 2',
    parentEntities: [],
  },
  {
    entityId: 3,
    entityType: 'dataset',
    id: 3,
    name: 'DATASET 1',
    parentEntities: [],
  },
  {
    entityId: 4,
    entityType: 'datafile',
    id: 4,
    name: 'DATAFILE 1',
    parentEntities: [],
  },
];

// Create our mocked datagateway-download settings file.
export const mockedSettings: Partial<DownloadSettings> = {
  facilityName: 'LILS',
  apiUrl: 'https://example.com/api',
  downloadApiUrl: 'https://example.com/downloadApi',
  idsUrl: 'https://example.com/ids',
  fileCountMax: 5000,
  totalSizeMax: 1000000000000,
  accessMethods: {
    https: {
      idsUrl: 'https://example.com/ids',
      displayName: 'HTTPS',
      description: 'Example description for <b>HTTPS</b> access method.',
    },
    globus: {
      idsUrl: 'https://example.com/ids',
      displayName: 'Globus',
      description: 'Example description for Globus access method.',
    },
  },
  uiFeatures: {
    downloadProgress: false,
  },
};
