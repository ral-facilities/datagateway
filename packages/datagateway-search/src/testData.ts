// Defines test data used for testing

import type { Investigation } from 'datagateway-common';

const mockInvestigation: Investigation = {
  id: 1,
  name: 'Test investigation name',
  title: 'Test investigation title',
  visitId: '1',
};

export { mockInvestigation };
