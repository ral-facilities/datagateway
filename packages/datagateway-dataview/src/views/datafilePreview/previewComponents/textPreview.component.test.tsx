import type { RenderResult } from '@testing-library/react';
import { render, screen, waitFor } from '@testing-library/react';
import { DGThemeProvider } from 'datagateway-common';
import * as React from 'react';
import { QueryClient, QueryClientProvider } from 'react-query';
import { Provider } from 'react-redux';
import { combineReducers, createStore, Store } from 'redux';
import { StateType } from '../../../state/app.types';
import dgdataviewReducer from '../../../state/reducers/dgdataview.reducer';
import {
  DecrementDatafilePreviewerZoomLevelType,
  IncrementDatafilePreviewerZoomLevelType,
} from '../state/actions';
import { mockDatafile, mockTxtFileContent } from '../testData';
import TextPreview from './textPreview.component';

function renderComponent(store: Store): RenderResult {
  return render(
    <DGThemeProvider>
      <Provider store={store}>
        <QueryClientProvider
          client={
            new QueryClient({ defaultOptions: { queries: { retry: false } } })
          }
        >
          <TextPreview
            datafile={mockDatafile}
            datafileContent={new Blob([mockTxtFileContent])}
          />
        </QueryClientProvider>
      </Provider>
    </DGThemeProvider>
  );
}

describe('TextPreview', () => {
  let store: Store;

  beforeEach(() => {
    store = createStore(
      combineReducers<Partial<StateType>>({
        dgdataview: dgdataviewReducer,
      })
    );
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it('should render given text file', async () => {
    const { asFragment } = renderComponent(store);
    // wait for the preview to finish loading
    await waitFor(() => {
      expect(
        screen.queryByText('datafiles.preview.txt.reading_content')
      ).toBeNull();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('should show loading when reading the text file from blob', () => {
    jest.spyOn(global, 'Blob').mockImplementationOnce(() => ({
      size: 123,
      type: 'text/plain',
      arrayBuffer: jest.fn(),
      slice: jest.fn(),
      stream: jest.fn(),
      text: () =>
        new Promise<string>(() => {
          // a promise that never resolves to pretend this is reading content
        }),
    }));

    renderComponent(store);

    expect(
      screen.getByText('datafiles.preview.txt.reading_content')
    ).toBeInTheDocument();
  });

  it('should show an error message when the blob cannot be read as text properly', async () => {
    jest.spyOn(global, 'Blob').mockImplementationOnce(() => ({
      size: 123,
      type: 'text/plain',
      arrayBuffer: jest.fn(),
      slice: jest.fn(),
      stream: jest.fn(),
      text: () => Promise.reject(),
    }));

    renderComponent(store);

    expect(
      await screen.findByText('datafiles.preview.txt.cannot_read_content')
    ).toBeInTheDocument();
  });

  describe('when the zoom level of datafile previewer changes', () => {
    it('should increase the font size if the zoom is increased', async () => {
      renderComponent(store);

      await waitFor(
        () =>
          screen.queryByText('datafiles.preview.txt.reading_content') === null
      );

      // increase the zoom level
      store.dispatch({
        type: IncrementDatafilePreviewerZoomLevelType,
      });

      expect(
        await screen.findByLabelText('datafiles.preview.txt.file_content_label')
      ).toHaveStyle({
        'font-size': '13px',
      });
    });

    it('should decrease the font size if the zoom is decreased', async () => {
      renderComponent(store);

      await waitFor(
        () =>
          screen.queryByText('datafiles.preview.txt.reading_content') === null
      );

      // increase the zoom level
      store.dispatch({
        type: DecrementDatafilePreviewerZoomLevelType,
      });

      expect(
        await screen.findByLabelText('datafiles.preview.txt.file_content_label')
      ).toHaveStyle({
        'font-size': '11px',
      });
    });
  });
});
