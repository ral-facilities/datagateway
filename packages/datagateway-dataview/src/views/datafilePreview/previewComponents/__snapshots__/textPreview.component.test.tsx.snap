// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextPreview should render given text file 1`] = `
<DocumentFragment>
  <pre
    aria-label="datafiles.preview.txt.file_content_label"
    class="css-h4y87t"
    font-size="12"
  >
    <span
      class="css-ppgvqg"
    >
      1
    </span>
    <code
      class="css-a948ge"
    >
      First line

    </code>
    <span
      class="css-ppgvqg"
    >
      2
    </span>
    <code
      class="css-a948ge"
    >
      Second line

    </code>
    <span
      class="css-ppgvqg"
    >
      3
    </span>
    <code
      class="css-a948ge"
    >
      Third line

    </code>
    <span
      class="css-ppgvqg"
    >
      4
    </span>
    <code
      class="css-a948ge"
    >
      

    </code>
  </pre>
</DocumentFragment>
`;
